clone/cmd/ctclone/ctclone.go

```diff
@@ -177,27 +178,37 @@ type reporter struct {
 	lastWorked   uint64
 	treeSize     uint64
 	epochWorked  time.Duration
+
+	mu sync.RWMutex
 }
 
 func (r *reporter) report() {
-	elapsed := time.Since(r.lastReport)
-	workRatio := r.epochWorked.Seconds() / elapsed.Seconds()
+	func() {
+		r.mu.RLock()
+		defer r.mu.RUnlock()
+		elapsed := time.Since(r.lastReport)
+		workRatio := r.epochWorked.Seconds() / elapsed.Seconds()
 
-	remaining := r.treeSize - r.lastReported - 1
-	rate := float64(r.lastWorked-r.lastReported) / elapsed.Seconds()
-	eta := time.Duration(float64(remaining)/rate) * time.Second
-	glog.Infof("%.1f leaves/s, last leaf=%d (remaining: %d, ETA: %s), time working=%.1f%%", rate, r.lastReported, remaining, eta, 100*workRatio)
+		remaining := r.treeSize - r.lastReported - 1
+		rate := float64(r.lastWorked-r.lastReported) / elapsed.Seconds()
+		eta := time.Duration(float64(remaining)/rate) * time.Second
+		glog.Infof("%.1f leaves/s, last leaf=%d (remaining: %d, ETA: %s), time working=%.1f%%", rate, r.lastReported, remaining, eta, 100*workRatio)
```

Comment:
Just a thought - you could make a snapshot copy of the values under lock and then release that so that you're not locked for the duration of the logging i/o on this line?

maybe something like:
```go
...

lw, lr, ew := func() (uint64, uint64, time.Duration) {
  r.mu.RLock()
  defer r.mu.RUnlock()
  return r.lastWorked, r.lastReported, r.epochWorked
}

// do calculations and log using local vars from above

// lock and update stats
}
```

Reply:
Nice. I've taken this on, and reconsidered what the lock was actually protecting. Instead of 3 lock blocks, there are now only 2 which also ensures work can't get missed, which was possible in the previous implementation (epochWorked could be read, then incremented, then zeroed).